internal/rotate.go

```diff
@@ -0,0 +1,88 @@
+package internal
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+)
+
+type CertRotationState struct {
+	EstServer string
+}
+
+type CertRotationHandler struct {
+	State           CertRotationState
+	stateFile       string
+	stateFileExists bool
+	app             *App
+	client          *http.Client
+	crypto          *EciesCrypto
+	steps           []CertRotationStep
+}
+
+type CertRotationStep interface {
+	Name() string
+	Completed(state CertRotationState) bool
+	Execute(handler *CertRotationHandler) error
+}
+
+func NewCertRotationHandler(app *App, stateFile, estServer string) CertRotationHandler {
+	client, crypto := createClient(app.sota)
+	handler := CertRotationHandler{
+		State:           CertRotationState{EstServer: estServer},
+		stateFile:       stateFile,
+		stateFileExists: true,
+		app:             app,
+		client:          client,
+		crypto:          crypto.(*EciesCrypto),
+		steps:           []CertRotationStep{},
+	}
+	bytes, err := os.ReadFile(stateFile)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			handler.stateFileExists = false
+		} else {
+			// Looks like we started a rotation, we should try and finish it
+			log.Printf("Error reading %s, return empty rotation state: %s", stateFile, err)
```

Comment:
it looks like a critical error, e.g. read error because of file permissions.

Reply:
its not a general purpose library, so we can just make this decision for ourselves. I think in the case we need to rotate.